### ISSUE TITLE
Several addons are not doing broadcast packets correctly

### ISSUE THREAD

**Issue by caffineehacker** [openhab/openhab2-addons#3054]
[Several addons](https://github.com/openhab/openhab2-addons/search?utf8=%E2%9C%93&q=%22255.255.255.255%22&type=) are not doing discovery via broadcast packets correctly. They hardcode 255.255.255.255 which will not work on all networks. Since the method of getting the correct broadcast addresses is non-trivial to get right I'd propose putting it in a common location for use by addons needing to use broadcast packets.

Where would be the preferred location for this common code? Would it be preferred that the addon simply provides a packet contents and a callback function and the rest is handled by common code or would it be preferred that the common code just gets the broadcast addresses?

For reference, here is the code I came up with to find the addresses:

```java
    private List<InetAddress> getBroadcastAddresses() {
        ArrayList<InetAddress> addresses = new ArrayList<>();

        System.setProperty("java.net.preferIPv4Stack", "true");

        try {
            for (NetworkInterface iface : Collections.list(NetworkInterface.getNetworkInterfaces())) {
                try {
                    if (iface.isLoopback()) {
                        logger.debug("Found loopback iface {}", iface.getName());
                        continue;
                    }

                    logger.debug("Found non-loopback iface {} with {} addresses", iface.getName(),
                            iface.getInterfaceAddresses().size());

                    for (InterfaceAddress address : iface.getInterfaceAddresses()) {
                        InetAddress broadcastAddress = address.getBroadcast();
                        if (broadcastAddress != null) {
                            logger.debug("Found broadcast address {} for iface {}", address.getBroadcast().toString(),
                                    iface.getName());
                            addresses.add(address.getBroadcast());
                        }
                    }
                } catch (SocketException e) {
                    logger.debug("Got exception in getBroadcastAddress", e);
                }
            }
        } catch (SocketException e) {
            logger.debug("Got exception in getBroadcastAddress", e);
        }

        try {
            // Add the default broadcast address as a fallback.
            addresses.add(InetAddress.getByName("255.255.255.255"));
        } catch (UnknownHostException e) {
        }
        return addresses;
    }
```

**Comment by kaikreuzer** [comment]
Thanks @caffineehacker for pointing this out!
Yes, such code is tricky and should be only done once in the framework and then be reused.
We have recently moved such helper methods to [NetUtil in ESH](https://github.com/eclipse/smarthome/blob/d43d4a0b3b37d880854689320da7a11c81fd1ddd/bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/net/NetUtil.java#L161-L198). I am not sure if this is already doing what you need or if some additional methods should be introduced - but at least that should be the right place to go for!

**Comment by caffineehacker** [comment]
Thanks, that's just what I was looking for. I'm going to clean up the other addons I found and send a PR.

**Comment by martinvw** [comment]
I assume this one can be closed now, given your PR is merged?

**Comment by mherwege** [comment]
@caffineehacker I am afraid we run into an issue with your solution. I agree the hard coded broadcast address is not appropriate. That’s why I added the broadcast methods in NetUtils, intention was to later call these in my Niko Home Control  binding. However, the findBroadcastAddress method is incomplete. It will just return the first broadcast address it finds, and that is random. It should probably return the broadcast address of the subnet that contains the primary ip address if it is provided. Otherwise we risk broadcasting to the wrong interface. This will be the case in my Niko Home Control setup at my place (a virtual VPN interface is always picked as first). I never got around fixing this.

**Comment by martinvw** [comment]
@caffineehacker @mherwege should I revert?

**Comment by kaikreuzer** [comment]
Or someone quickly creates a PR against the NetUtil implementation?

**Comment by caffineehacker** [comment]
I would do a rollback since there are some things outside of NetUtil that will probably need to change as well given this information. I'll send a new PR after the rollback (probably not until later in the week) that also fixes up the places that used a single broadcast address to use the getAllBroadcastAddresses() (as an aside, we might want to consider getting rid of getBroadcastAddress() since it will always be deficient in some way compared to getAllBroadcastAddresses()). Once that is done, I'll consider making a function in NetUtil that does all of the broadcast logic for you, including listening and calling a callback with the responses since that seems to be another place some addons are having issues (a few only listen for the first response and therefore won't be able to find multiple devices).

**Comment by mherwege** [comment]
@caffineehacker Did you progress on this? I have made a change to the NetUtil package, so it will return the broadcast address of the interface containing the primary IP in a new method (and thereby deprecating getBroadcastAddress()). I also prepared a patch for the NikoHomeControl binding to use that. This approach avoids broadcasting on all subnets on a multi-homed system and has the least impact in my case.
I think it would then be up to the binding developers to decide if they want to do broadcasts on multiple interfaces (iterating over the list of broadcast addresses) or just the primary one.
Unless you have something nicer up your sleeves, I am happy to create pull requests.

**Comment by caffineehacker** [comment]
Looks like Kai reverted my change (thanks!) so we're at least not breaking anyone at the moment. I don't know if it makes sense to have each and every addon decide if they want to broadcast on a single interface or all the interfaces as that would just leave a lot of inconsistency. In particular, anyone with multiple NICs might have a hard time using some addons while others would work fine. I'd push to have all addons iterate over the broadcast addresses and broadcast widely.

That said, since the change has been made to the NetUtil package we can probably roll forward again. I'll create a pull release (probably not until next weekend) that incorporates my original changes unless there are objections.

**Comment by triller-telekom** [comment]
With PR #4848 we would have `getConfiguredBroadcastAddress()` in the `NetworkAddressService` that returns the broadcast address where the configured primary IP is in. Additionally you can choose a different broadcast address in another network via paper UI.

@caffineehacker What are you missing in this solution? Do you need to configure *multiple* broadcast addresses? If so, do you have a real world use-case for that?

**Comment by caffineehacker** [comment]
I don't have a personal use case that would require multiple broadcast addresses, just a theoretical case where the user has two subnets on their network that they need to broadcast to (or two network interfaces attached to the board such as would be the case if the user want to isolate their home automation network from their internet network and only have the openHab machine be in between). That said, these should be fairly uncommon examples so the current utility function should be good enough (also note that we could always put the standard UDP broadcast and handle code in a utility function which would make updating this kind of behavior much easier in the future if we do decide we want more than one network interface to be supported).
